_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+If there is any activity, then `ActivityTrackingInputStream` and `ActivityTrackingOutputStream` decorators will reset the counter.
+
+This solution is not ideal, because we have to trust the client of this class to run the `isInactive()` method with the `checkKeepAliveInterval` interval. On the other hand, the `Connection` class won’t be visible to the end user of our `SocketClient` so I think we can live with that.
+
+#### Connections
+Now we need a container from which we can lease connections.
+``` kotlin
+private class Connections {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(Connections::class.java)
+    }
+
+    private val leased = ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>()
+    private val available = ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>()
+
+    fun addLeased(connection: Connection) { // (2)
+        leased(connection.destination) += connection
+    }
+
+    fun lease(destination: ConnectionDestination): Connection? { // (1)
+        // There can be already closed connection, but we want to return an active one
+        do {
+            val connection = available(destination).poll()
+            if (connection == null) {
+                return null
+            } else if (!connection.isClosed()) {
+                addLeased(connection)
+                return connection
+            }
+        } while(true)
+    }
+
+    fun release(connection: Connection) { // (3)
+        leased(connection.destination).remove(connection)
+        available(connection.destination) += connection
+    }
+
+    fun closeStale() { // (4)
+        closeStale(leased)
+        closeStale(available)
+    }
+
+    private fun leased(destination: ConnectionDestination) =
+            leased.computeIfAbsent(destination) { LinkedBlockingDeque() }
+
+    private fun available(destination: ConnectionDestination) =
+            available.computeIfAbsent(destination) { LinkedBlockingDeque() }
+
+    private fun closeStale(
+            connectionsForDestination: ConcurrentHashMap<ConnectionDestination, BlockingDeque<Connection>>) {
+        connectionsForDestination.forEach { _, connections ->
+            val staleConnections = connections.filter(Connection::isInactive)
+            connections.removeAll(staleConnections)
+            staleConnections.forEach {
+                logger.debug("Closing connection to ${it.destination} due to not being active")
+                it.close()
+            }
+        }
+    }
+
+}
+```
+We keep separate queues of connections for leased and available connections. We can lease a connection (1), but we have to be careful not to return a closed connection. We can add already leased connection (2). This will be useful when we will create a connection and return it immediately to the client. We can release connection (3) so it is placed back into the leased queue.
```

Comment:
`This will be useful when we will create a connection and return it immediately to the client` => `This will be useful when we create a connection and return it immediately to the client`

"There is no general rule against using "will" twice in the same sentence. However, your sentence is incorrect because adverbial clauses with when do not usually use the modal will. To indicate future time within a when-clause, use the simple present:" https://english.stackexchange.com/questions/41123/using-will-twice-in-a-sentence